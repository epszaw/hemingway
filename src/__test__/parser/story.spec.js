const path = require('path')
const { StoryParser } = require('src/parser')
const { readFile } = require('src/utils/fs')
const parsedCommonStory = require('__fixtures__/parsed/common_story')
const parsedMultipleStories = require('__fixtures__/parsed/multiple_stories')

const readRawStoryFixture = async name => {
  const res = await readFile(
    path.resolve(__dirname, `../__fixtures__/raw/${name}.md`),
    'utf8'
  )

  return res
}

describe('StoryParser', () => {
  let singleStory
  let multipleStories
  let untitledStory
  let titledStory

  beforeAll(async () => {
    singleStory = await readRawStoryFixture('common_story')
    multipleStories = await readRawStoryFixture('multiple_stories')
    untitledStory = await readRawStoryFixture('untitled_story')
    titledStory = await readRawStoryFixture('titled_story')
  })

  describe('StoryParser – signStoriesNames', () => {
    it('should add filename as story title if title is not exist', () => {
      expect(
        StoryParser.signStoriesNames([
          {
            filename: 'untitled_story.md',
            source: untitledStory,
          },
        ])
      ).toEqual([titledStory])
    })
  })

  describe('StoryParser – parseStory', () => {
    it('should correctly parse story without annotations', () => {
      const res = StoryParser.parseStory(singleStory)

      expect(res).toEqual(parsedCommonStory)
    })
  })

  describe('StoryParser – parseStories', () => {
    it('should correctly parse multiple stories without annotations', () => {
      const res = StoryParser.parseStories([
        {
          filename: 'stories.md',
          source: multipleStories,
        },
      ])

      expect(res).toEqual(parsedMultipleStories)
    })
  })
})
