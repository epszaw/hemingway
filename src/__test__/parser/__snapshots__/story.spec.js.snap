// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoryParser StoryParser – parseStories should correctly parse multiple stories without annotations 1`] = `
Array [
  Object {
    "actions": Array [
      Object {
        "commands": Array [
          Object {
            "args": Array [
              "https://google.com",
            ],
            "modifier": null,
            "name": "OPEN",
          },
        ],
        "def": "OPEN \\"https://google.com\\"",
      },
      Object {
        "commands": Array [
          Object {
            "args": Array [
              "#lst-ib",
            ],
            "modifier": null,
            "name": "FIND",
          },
          Object {
            "args": Array [
              "search-input",
            ],
            "modifier": null,
            "name": "AS",
          },
        ],
        "def": "FIND \\"#lst-ib\\" AS \\"search-input\\"",
      },
      Object {
        "commands": Array [
          Object {
            "args": Array [
              "search-input",
            ],
            "modifier": null,
            "name": "TAKE",
          },
          Object {
            "args": Array [
              "Hello{Enter}",
            ],
            "modifier": null,
            "name": "TYPE",
          },
        ],
        "def": "TAKE \\"search-input\\" TYPE \\"Hello{Enter}\\"",
      },
      Object {
        "commands": Array [
          Object {
            "args": Array [
              "#rhscol",
            ],
            "modifier": null,
            "name": "WAIT",
          },
        ],
        "def": "WAIT \\"#rhscol\\"",
      },
      Object {
        "commands": Array [
          Object {
            "args": Array [
              ".bkWMgd",
            ],
            "modifier": null,
            "name": "FIND",
          },
          Object {
            "args": Array [
              "length",
            ],
            "modifier": null,
            "name": "HAVE",
          },
          Object {
            "args": Array [
              "0",
            ],
            "modifier": "NOT",
            "name": "EQUAL",
          },
        ],
        "def": "FIND \\".bkWMgd\\" HAVE \\"length\\" NOT EQUAL \\"0\\"",
      },
    ],
    "name": "Story name 1",
  },
  Object {
    "actions": Array [
      Object {
        "commands": Array [
          Object {
            "args": Array [
              "https://google.com",
            ],
            "modifier": null,
            "name": "OPEN",
          },
        ],
        "def": "OPEN \\"https://google.com\\"",
      },
      Object {
        "commands": Array [
          Object {
            "args": Array [
              "#lst-ib",
            ],
            "modifier": null,
            "name": "FIND",
          },
          Object {
            "args": Array [
              "search-input",
            ],
            "modifier": null,
            "name": "AS",
          },
        ],
        "def": "FIND \\"#lst-ib\\" AS \\"search-input\\"",
      },
      Object {
        "commands": Array [
          Object {
            "args": Array [
              "search-input",
            ],
            "modifier": null,
            "name": "TAKE",
          },
          Object {
            "args": Array [
              "Hello{Enter}",
            ],
            "modifier": null,
            "name": "TYPE",
          },
        ],
        "def": "TAKE \\"search-input\\" TYPE \\"Hello{Enter}\\"",
      },
      Object {
        "commands": Array [
          Object {
            "args": Array [
              "#rhscol",
            ],
            "modifier": null,
            "name": "WAIT",
          },
        ],
        "def": "WAIT \\"#rhscol\\"",
      },
      Object {
        "commands": Array [
          Object {
            "args": Array [
              ".bkWMgd",
            ],
            "modifier": null,
            "name": "FIND",
          },
          Object {
            "args": Array [
              "length",
            ],
            "modifier": null,
            "name": "HAVE",
          },
          Object {
            "args": Array [
              "0",
            ],
            "modifier": "NOT",
            "name": "EQUAL",
          },
        ],
        "def": "FIND \\".bkWMgd\\" HAVE \\"length\\" NOT EQUAL \\"0\\"",
      },
    ],
    "name": "Story name 2",
  },
  Object {
    "actions": Array [
      Object {
        "commands": Array [
          Object {
            "args": Array [
              "https://google.com",
            ],
            "modifier": null,
            "name": "OPEN",
          },
        ],
        "def": "OPEN \\"https://google.com\\"",
      },
      Object {
        "commands": Array [
          Object {
            "args": Array [
              "#lst-ib",
            ],
            "modifier": null,
            "name": "FIND",
          },
          Object {
            "args": Array [
              "search-input",
            ],
            "modifier": null,
            "name": "AS",
          },
        ],
        "def": "FIND \\"#lst-ib\\" AS \\"search-input\\"",
      },
      Object {
        "commands": Array [
          Object {
            "args": Array [
              "search-input",
            ],
            "modifier": null,
            "name": "TAKE",
          },
          Object {
            "args": Array [
              "Hello{Enter}",
            ],
            "modifier": null,
            "name": "TYPE",
          },
        ],
        "def": "TAKE \\"search-input\\" TYPE \\"Hello{Enter}\\"",
      },
      Object {
        "commands": Array [
          Object {
            "args": Array [
              "#rhscol",
            ],
            "modifier": null,
            "name": "WAIT",
          },
        ],
        "def": "WAIT \\"#rhscol\\"",
      },
      Object {
        "commands": Array [
          Object {
            "args": Array [
              ".bkWMgd",
            ],
            "modifier": null,
            "name": "FIND",
          },
          Object {
            "args": Array [
              "length",
            ],
            "modifier": null,
            "name": "HAVE",
          },
          Object {
            "args": Array [
              "0",
            ],
            "modifier": "NOT",
            "name": "EQUAL",
          },
        ],
        "def": "FIND \\".bkWMgd\\" HAVE \\"length\\" NOT EQUAL \\"0\\"",
      },
    ],
    "name": "Story name 3",
  },
]
`;

exports[`StoryParser StoryParser – parseStory should correctly parse actions with separated arguments 1`] = `
Object {
  "actions": Array [
    Object {
      "commands": Array [
        Object {
          "args": Array [
            ".foo",
          ],
          "modifier": null,
          "name": "FIND",
        },
        Object {
          "args": Array [
            "length",
          ],
          "modifier": null,
          "name": "HAVE",
        },
        Object {
          "args": Array [
            "0",
          ],
          "modifier": "NOT",
          "name": "EQUAL",
        },
      ],
      "def": "FIND \\".foo\\" HAVE \\"length\\" NOT EQUAL \\"0\\"",
    },
    Object {
      "commands": Array [
        Object {
          "args": Array [
            ".foo .bar",
          ],
          "modifier": null,
          "name": "FIND",
        },
        Object {
          "args": Array [
            "length",
          ],
          "modifier": null,
          "name": "HAVE",
        },
        Object {
          "args": Array [
            "0",
          ],
          "modifier": "NOT",
          "name": "EQUAL",
        },
      ],
      "def": "FIND \\".foo .bar\\" HAVE \\"length\\" NOT EQUAL \\"0\\"",
    },
    Object {
      "commands": Array [
        Object {
          "args": Array [
            "[data-test='foo']",
          ],
          "modifier": null,
          "name": "FIND",
        },
        Object {
          "args": Array [
            "length",
          ],
          "modifier": null,
          "name": "HAVE",
        },
        Object {
          "args": Array [
            "0",
          ],
          "modifier": "NOT",
          "name": "EQUAL",
        },
      ],
      "def": "FIND \\"[data-test='foo']\\" HAVE \\"length\\" NOT EQUAL \\"0\\"",
    },
    Object {
      "commands": Array [
        Object {
          "args": Array [
            "[data-test='foo'] .bar",
          ],
          "modifier": null,
          "name": "FIND",
        },
        Object {
          "args": Array [
            "length",
          ],
          "modifier": null,
          "name": "HAVE",
        },
        Object {
          "args": Array [
            "0",
          ],
          "modifier": "NOT",
          "name": "EQUAL",
        },
      ],
      "def": "FIND \\"[data-test='foo'] .bar\\" HAVE \\"length\\" NOT EQUAL \\"0\\"",
    },
    Object {
      "commands": Array [
        Object {
          "args": Array [
            ".foo [data-test='bar']",
          ],
          "modifier": null,
          "name": "FIND",
        },
        Object {
          "args": Array [
            "length",
          ],
          "modifier": null,
          "name": "HAVE",
        },
        Object {
          "args": Array [
            "0",
          ],
          "modifier": "NOT",
          "name": "EQUAL",
        },
      ],
      "def": "FIND \\".foo [data-test='bar']\\" HAVE \\"length\\" NOT EQUAL \\"0\\"",
    },
  ],
  "name": "Arguments",
}
`;

exports[`StoryParser StoryParser – parseStory should correctly parse story with diffirent operators 1`] = `
Object {
  "actions": Array [
    Object {
      "commands": Array [
        Object {
          "args": Array [
            ".foo button.bar",
          ],
          "modifier": null,
          "name": "FIND",
        },
        Object {
          "args": Array [],
          "modifier": null,
          "name": "CLICK",
        },
      ],
      "def": "FIND \\".foo button.bar\\" CLICK",
    },
  ],
  "name": "Operators",
}
`;

exports[`StoryParser StoryParser – parseStory should correctly parse story without annotations 1`] = `
Object {
  "actions": Array [
    Object {
      "commands": Array [
        Object {
          "args": Array [
            "https://google.com",
          ],
          "modifier": null,
          "name": "OPEN",
        },
      ],
      "def": "OPEN \\"https://google.com\\"",
    },
    Object {
      "commands": Array [
        Object {
          "args": Array [
            "#lst-ib",
          ],
          "modifier": null,
          "name": "FIND",
        },
        Object {
          "args": Array [
            "search-input",
          ],
          "modifier": null,
          "name": "AS",
        },
      ],
      "def": "FIND \\"#lst-ib\\" AS \\"search-input\\"",
    },
    Object {
      "commands": Array [
        Object {
          "args": Array [
            "search-input",
          ],
          "modifier": null,
          "name": "TAKE",
        },
        Object {
          "args": Array [
            "Hello{Enter}",
          ],
          "modifier": null,
          "name": "TYPE",
        },
      ],
      "def": "TAKE \\"search-input\\" TYPE \\"Hello{Enter}\\"",
    },
    Object {
      "commands": Array [
        Object {
          "args": Array [
            "#rhscol",
          ],
          "modifier": null,
          "name": "WAIT",
        },
      ],
      "def": "WAIT \\"#rhscol\\"",
    },
    Object {
      "commands": Array [
        Object {
          "args": Array [
            ".bkWMgd",
          ],
          "modifier": null,
          "name": "FIND",
        },
        Object {
          "args": Array [
            "length",
          ],
          "modifier": null,
          "name": "HAVE",
        },
        Object {
          "args": Array [
            "0",
          ],
          "modifier": "NOT",
          "name": "EQUAL",
        },
      ],
      "def": "FIND \\".bkWMgd\\" HAVE \\"length\\" NOT EQUAL \\"0\\"",
    },
  ],
  "name": "Story name",
}
`;
